### ISSUE TITLE
Windows Port (Using vJoy)

### ISSUE THREAD

**Issue by dantarion** [altf4/SmashBot#69]
I will eventually port your bot to windows Dolphin in order to pit it against mine.

In order to do so I will have to change the code that relies on the input pipe to use vJoy (There is a C++ lib for pushing data to a virtual joystick, I use the C# one with my MarthBot)

The other thing I will have to figure out is how to port your memory reading code.
I use ReadProcessMemory on Windows, so I will have to use that to replace the MemoryWatcher method being used by SmashBot.

Hopefully I can get this working :(
If you have any input or information that can help with this, please let me know :D

**Comment by altf4** [comment]
That is probably not the best way to approach it. 

Dolphin already has all the features that you need, they are just not all present on the Dolphin windows version. Particularly named pipe controller inputs and MemoryWatcher. Those are really the only two things that are missing for a windows version to work. The SmashBot code would need very little other changes.

Named pipe inputs are a way of programmatically pressing buttons on a virtual controller directly into Dolphin. I tried many options before coming to the conclusion that this way necessary, and I'd recommend highly against going another direction. SmashBot relies heavily on the fact that there is zero latency between button press and engine processing. HID systems (keyboard / mouse / joystick) are invariably buffered input with a potentially arbitrary amount of latency. 

MemoryWatcher is also integral to how SmashBot works in that it gets us reliable memory dumps of exactly what we need from the internals of the game. It is a feature in Dolphin that allows us to query for memory regions and get updates when those memory regions change. Trying to read the raw memory of the dolphin process through ReadProcessMemory would be horrible and unreliable. MemoryWatcher talks over unix domain sockets back to the outside process, so that won't work on Windows. But there is likely an alternative there.

If I were you, I would talk to the Dolphin developers, particularly @spxtr about these. I'd help, but I'm not really a windows guy, and don't even have a windows machine around to test on.

That said, I really look forward to the day when we can have AI tournaments. I think that would be really fun!

**Comment by spxtr** [comment]
I would also suggest porting the dolphin changes to Windows. I don't know much about Windows IPC but I think there are analogues of named pipes and unix sockets that should work.

**Comment by vladfi1** [comment]
In my fork of dolphin I changed the memory watcher to use zmq instead of
named pipes, which should work on windows. Using zmq also allows proper
synchronization between dolphin and the bot.

**Comment by dantarion** [comment]
Interesting. From what I have seen from my testing (I am only a few days of development in and I just wanted to see how you guys were rigging things up, so I haven't gotten into optimization yet...) everything seems really consistant although ill admit I my inputs seem to delay a frame every now and then.

I will look into a way of doing named pipes/sockets on Windows with Dolphin, I came up with my method because I didn't want to modify dolphin. I am aiming to get to a point where at some point, all of our bots can play each other, regardless of platform.

**Comment by spxtr** [comment]
> ill admit I my inputs seem to delay a frame every now and then.

I'm not sure what causes that, but I've noticed it and so has @altf4 I think. From my testing it actually seems like melee skips a frame, or maybe dolphin gets behind and catches up or something? I never figured it out.

> I am aiming to get to a point where at some point, all of our bots can play each other, regardless of platform.

We should have a tournament!

**Comment by dantarion** [comment]
Could it be this?
https://www.reddit.com/r/smashbros/comments/4pcfbv/melee_on_console_how_much_input_lag_is_there_and/

It would definitely affect bots moreso than actual players, as we are often trying to do things on the first actionable frame repeatedly.

**Comment by altf4** [comment]
Yes, that issue has been all figured out. 

It's a bug in Melee itself, and Dan Salvato made a patch via Gecko Code for it. It's included in the Melee netplay community settings in Dolphin 5.0. So just apply that Gecko code and it will fix the problem.

As for AI tournaments... we'd have to come up with some ground rules about what constitutes "cheating". For instance, there are Starcraft AI tournaments, and the community came together to decide on what cheating means. (No map hacking, for instance)

**Comment by spxtr** [comment]
Oh awesome, I didn't know that was fixed.

For AI tournaments I think the cheating rules could be pretty simple: you both use the same MemoryWatcher and you both write to a named pipe (or whatever the equivalent is on windows) and you don't hog the cpu.

However, things like character choice and the like could be tricky. smashbot only plays fox vs marth, and I'm not sure about the others.

**Comment by altf4** [comment]
I'm work on supporting other characters and stages now, actually. SmashBot will play **as** Fox for the indefinite future, but other matchups will be coming shortly.

Other questions to ask are stage settings: I'd like to use frozen stages to reduce complexity. IE: No Pokemon Stadium transformations.

**Comment by dantarion** [comment]
https://msdn.microsoft.com/en-us/library/windows/desktop/aa365590(v=vs.85).aspx
Looks like windows straight up supports named pipes after all?
I guess I will start the fun process of getting code accepted into Dolphin!
This should be directly compatible with the way you guys are currently reading memory/sending inputs to dolphin, right?

**Comment by altf4** [comment]
Yea, that should work just fine. When you do, use the Locations.txt from the SmashBot code. It has all the necessary game state info to play. (Except for projectile info, which I'm working on and should be coming soon)

Notably, we should agree on what information we're reading from the game. But I think you'll find in SmashBot's gamestate info sufficient but not over-reaching. (It doesn't do things like read the opponent's controller state, or predict the values of "random" numbers)

**Comment by dantarion** [comment]
The only thing that makes me wonder about the MemoryWatcher is if anything if dynamically allocated in a way that the current Locations.txt format can't exactly handle, such as reading an array of platform locations, projectiles, hurtboxes, hitboxes, etc. Anyways, this is out of scope of what I originally created this issue for!

I will aim to eventually add proper support for the missing features to Windows Dolphin, and work with you guys towards establishing a common config so that its easy to load up bots at the same time and have them battle it out!

**Comment by altf4** [comment]
MemoryWatcher has support for following pointers, which is how it currently accesses character info. The game dynamically allocates a struct with some info. The address of the pointer is in a stable memory location that doesn't change, but the location on the heap of its data is dynamic. That's why we can follow pointers. 

The only thing it can't do just yet is follow linked lists. (which is how projectiles are stored) So I'll need to add support for that before we can read where projectiles are.

**Comment by altf4** [comment]
Oh, also in SmashBot you'll want to look at the code here:

https://github.com/altf4/SmashBot/blob/master/Util/Controller.cpp#L30

SmashBot automatically tries to create a named pipe file in a location that Dolphin will need it. That location will need to be changed on Windows. (You may need to add in some macro preprocessor defines to remove some of the Unix-specific path functions)

Same thing for MemoryWatcher here:

https://github.com/altf4/SmashBot/blob/master/Util/MemoryWatcher.cpp#L18

But you probably already noticed that one.

Those **should** be the only Unix-specific pieces of code in the codebase. Otherwise, a Windows port ought to be simple.

**Comment by spxtr** [comment]
@dantarion Thanks, that sounds awesome! Feel free to bother me here or in #dolphin-dev on freenode, though I may not respond right away.

MemoryWatcher is pretty unintelligent right now. It doesn't even validate the memory address it's about to look up. There was some work involving making a more powerful dolphin debug controller (https://github.com/dolphin-emu/dolphin/pull/3031), though it didn't go very far. Might be worth looking at.

**Comment by altf4** [comment]
Closed for v2.0 (look at libmelee for windows port talk)
